server/import_certify.go

```diff
@@ -0,0 +1,188 @@
+package server
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/subtle"
+	"errors"
+	"fmt"
+
+	rpb "github.com/google/go-tpm-tools/proto/register_credential"
+	tpb "github.com/google/go-tpm-tools/proto/tpm"
+	"github.com/google/go-tpm/tpm2"
+)
+
+// this file aims to implement the verifier side of https://trustedcomputinggroup.org/wp-content/uploads/EK-Based-Key-Attestation-with-TPM-Firmware-Version-V1-RC1_9July2025.pdf#page=8
+// for reference: https://github.com/TrustedComputingGroup/tpm-fw-attestation-reference-code
+
+var (
+	errCertifiedWrongName = errors.New("incorrect name")
+	errWrongHashAlg       = errors.New("wrong hash algorithm")
+	errInvalidHMAC        = errors.New("invalid HMAC")
+	errInvalidAttestation = errors.New("attestation statement was invalid")
+)
+
+// CreateRestrictedHMACBlob generates a new HMAC key and wraps it to the given EK.
+func CreateRestrictedHMACBlob(tPublic *tpm2.TPMTPublic) (*tpb.ImportBlob, []byte, error) {
+	encap, err := tpm2.ImportEncapsulationKey(tPublic)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	hmacKey := make([]byte, 32)
+	pub, sensitive := generateRestrictedHMACKey(hmacKey)
+	name, err := tpm2.ObjectName(pub)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	duplicate, inSymSeed, err := tpm2.CreateDuplicate(rand.Reader, encap, name.Buffer, tpm2.Marshal(sensitive))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &tpb.ImportBlob{
+		PublicArea:    tpm2.Marshal(pub),
+		Duplicate:     duplicate,
+		EncryptedSeed: inSymSeed,
+	}, hmacKey, nil
+}
+
+func VerifyCertifiedAKBlob(req *rpb.CertifiedBlob, secret []byte) error {
+	akPub, err := tpm2.Unmarshal[tpm2.TPMTPublic](req.GetAkPub())
+	if err != nil {
+		return err
+	}
+
+	akName, err := tpm2.ObjectName(akPub)
+	if err != nil {
+		return err
+	}
+
+	signature, err := tpm2.Unmarshal[tpm2.TPMTSignature](req.GetRawSig())
+	if err != nil {
+		return err
+	}
+
+	hmac, err := signature.Signature.HMAC()
+	if err != nil {
+		return err
+	}
+
+	if err := verifyHMAC(secret, req.GetCertifyInfo(), hmac); err != nil {
+		return err
+	}
+
+	attest, err := tpm2.Unmarshal[tpm2.TPMSAttest](req.GetCertifyInfo())
+	if err != nil {
+		return err
+	}
+
+	if err := verifyAttest(attest); err != nil {
+		return err
+	}
+
+	certify, err := attest.Attested.Certify()
+	if err != nil {
+		return err
+	}
+
+	if err := verifyCertify(akName, certify); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// generateRestrictedHMACKey writes a new hmac to the input parameter and produces the pub/priv tpm2 structures
+func generateRestrictedHMACKey(hmacKey []byte) (*tpm2.TPMTPublic, *tpm2.TPMTSensitive) {
+	// Generate the random obfuscation value and key
+	obfuscate := make([]byte, 32)
+	rand.Read(obfuscate)
+	rand.Read(hmacKey[:])
+
+	// Unique for a KEYEDHASH object is H_nameAlg(obfuscate | key)
+	// See Part 1, "Public Area Creation"
+	h := sha256.New()
+	h.Write(obfuscate)
+	h.Write(hmacKey[:])
+
+	pub := &tpm2.TPMTPublic{
```

Comment:
This is heavily duplicated with https://github.com/google/go-tpm-tools/blob/main/server/key_conversion.go#L52-L64. Can we extend that to pass in the correct Object attributes (sealedDataObjects leave it unset; imported HMAC keys should set)